In [1]:
import numpy as np
from sklearn import datasets
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
import time

In [2]:
fashion_mnist = datasets.fetch_openml('Fashion-MNIST', version=1)
X = fashion_mnist.data
y = fashion_mnist.target.astype(np.int)

X_linear = X[np.isin(y, [0, 1])]
y_linear = y[np.isin(y, [0, 1])]

X_nonlinear = X[np.isin(y, [0, 6])]
y_nonlinear = y[np.isin(y, [0, 6])]
y_nonlinear = np.where(y_nonlinear == 0, -1, 1)

/usr/local/lib/python3.10/dist-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(
<ipython-input-2-817c4bfa47a0>:3: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  y = fashion_mni

In [3]:
X_linear_sampled, _, y_linear_sampled, _ = train_test_split(X_linear, y_linear, train_size=0.1, random_state=42)

X_nonlinear_sampled, _, y_nonlinear_sampled, _ = train_test_split(X_nonlinear, y_nonlinear, train_size=0.1, random_state=42)

X_train_linear, X_test_linear, y_train_linear, y_test_linear = train_test_split(X_linear_sampled, y_linear_sampled, test_size=0.3, random_state=42)
X_train_nonlinear, X_test_nonlinear, y_train_nonlinear, y_test_nonlinear = train_test_split(X_nonlinear_sampled, y_nonlinear_sampled, test_size=0.3, random_state=42)


In [4]:
svm_hard_linear = SVC(kernel='linear', C=3)
svm_hard_nonlinear = SVC(kernel='linear', C=3)


In [5]:

print("Training on linearly separable dataset...")
start_time = time.time()
svm_hard_linear.fit(X_train_linear, y_train_linear)
end_time = time.time()
print(f"Training of linearly separable dataset completed. Time taken: {end_time - start_time:.2f} seconds")

print("Training on non-linearly separable dataset...")
start_time = time.time()
svm_hard_nonlinear.fit(X_train_nonlinear, y_train_nonlinear)
end_time = time.time()
print(f"Training of non-linearly separable dataset completed. Time taken: {end_time - start_time:.2f} seconds")

y_pred_linear = svm_hard_linear.predict(X_test_linear)
print("Linearly Separable Dataset:")
print(classification_report(y_test_linear, y_pred_linear))

y_pred_nonlinear = svm_hard_nonlinear.predict(X_test_nonlinear)
print("Non-Linearly Separable Dataset:")
print(classification_report(y_test_nonlinear, y_pred_nonlinear))


Training on linearly separable dataset...
Training of linearly separable dataset completed. Time taken: 0.04 seconds
Training on non-linearly separable dataset...
Training of non-linearly separable dataset completed. Time taken: 0.27 seconds
Linearly Separable Dataset:
              precision    recall  f1-score   support

           0       0.98      0.98      0.98       204
           1       0.98      0.98      0.98       216

    accuracy                           0.98       420
   macro avg       0.98      0.98      0.98       420
weighted avg       0.98      0.98      0.98       420

Non-Linearly Separable Dataset:
              precision    recall  f1-score   support

          -1       0.82      0.75      0.79       223
           1       0.75      0.82      0.78       197

    accuracy                           0.78       420
   macro avg       0.78      0.79      0.78       420
weighted avg       0.79      0.78      0.78       420



Results
Linearly Separable Dataset
Accuracy: 98%
Precision and Recall: Both approximately 98% for categories 0 and 1.
Non-Linearly Separable Dataset
Accuracy: 78%
Precision and Recall: Varied between 75% to 82%, with category 1 showing a slightly higher recall.

Discussion
Linearly Separable Dataset
The high performance of the hard margin SVM on the linearly separable dataset indicates that the model effectively found a hyperplane in the feature space that accurately separated the two categories. Due to the inherent linear separability of the dataset, the model was able to classify samples with high precision and recall.

Non-Linearly Separable Dataset
Although hard margin SVMs are typically not suitable for non-linearly separable datasets, in our case, it still showed relatively high accuracy. This might indicate that categories 0 and 6 have some degree of linear separability in the feature space, or the overlap in the dataset was not very significant. However, hard margin SVMs may face challenges in dealing with truly non-linearly separable data, as they require complete linear separability and are very sensitive to outliers.
The failing aspect of this method is that it requires an exceedingly long time to train, to the extent that I had to reduce the size of the dataset.

Limitations of Hard Margin SVM
The main limitation of the hard margin SVM in dealing with non-linearly separable data lies in its assumption of complete linear separability. In real-world datasets, this assumption is often unrealistic, as data often contains noise and outliers. Hard margin SVMs are highly sensitive to these outliers, which might lead to model overfitting or inability to find an effective decision boundary. For more complex or heavily overlapping data distributions, soft margin SVMs or SVMs with non-linear kernels might need to be considered.

Conclusion
Our experiments show that hard margin SVMs are highly effective in dealing with linearly separable datasets, but they may have limitations when handling non-linearly separable datasets. Although they performed well in our case, in practical applications, choosing the right model and parameters is crucial for optimal performance.

